ext/curl/interface.c

```diff
@@ -768,6 +774,61 @@ static size_t curl_xferinfo(void *clientp, curl_off_t dltotal, curl_off_t dlnow,
 /* }}} */
 #endif
 
+#if LIBCURL_VERSION_NUM >= 0x075400
+static int curl_ssh_hostkeyfunction(void *clientp, int keytype, const char *key, size_t keylen)
+{
+	php_curl *ch = (php_curl *)clientp;
+	php_curl_callback *t = ch->handlers.sshhostkey;
+	int rval = CURLKHMATCH_MISMATCH; /* cancel connection in case of an exception */
+
+#if PHP_CURL_DEBUG
+	fprintf(stderr, "curl_ssh_hostkeyfunction() called\n");
+	fprintf(stderr, "clientp = %x, keytype = %d, key = %s, keylen = %zu\n", clientp, keytype, key, keylen);
+#endif
+
+	zval argv[4];
+	zval retval;
+	zend_result error;
+	zend_fcall_info fci;
+
+	GC_ADDREF(&ch->std);
+	ZVAL_OBJ(&argv[0], &ch->std);
+	ZVAL_LONG(&argv[1], keytype);
+	ZVAL_STRING(&argv[2], key);
+	ZVAL_LONG(&argv[3], keylen);
+
+	fci.size = sizeof(fci);
+	ZVAL_COPY_VALUE(&fci.function_name, &t->func_name);
+	fci.object = NULL;
+	fci.retval = &retval;
+	fci.param_count = 4;
+	fci.params = argv;
+	fci.named_params = NULL;
+
+	ch->in_callback = 1;
+	error = zend_call_function(&fci, &t->fci_cache);
+	ch->in_callback = 0;
+	if (error == FAILURE) {
+		php_error_docref(NULL, E_WARNING, "Cannot call the CURLOPT_SSH_HOSTKEYFUNCTION");
```

Comment:
As I can see we already know the function pointer (as we have an FCC) so you probably can use the new API ``zend_call_known_fcc()`` which foregoes the creation of an FCI struct.

Reply:
Ah, thanks for pointing this out. I copied this code from other functions so probably the others should also be changed into the new API (converting the others should happen in another PR)

Reply:
I don't think I can use that function. If the invocation fails, `zend_call_known_fcc` produces a core error: `zend_error_noreturn(E_CORE_ERROR, "Couldn't execute method %s%s%s"` while the current implementations used in curl produce a warning. I don't know how the invocation can fail, maybe it's not really, or rarely possible in practice?

Reply:
No, you are correct... when setting up the callback cURL doesn't check that it is indeed callable. (See lines 2200 to 2207 of interface.c (on master)). If you add a test (inspired by ``curl_setopt_CURLOPT_READFUNCTION.phpt``):

```php
$oCurl = curl_init();
curl_setopt($oCurl, CURLOPT_URL,          $sWriteUrl);
curl_setopt($oCurl, CURLOPT_UPLOAD,       1);
curl_setopt($oCurl, CURLOPT_READFUNCTION, "non_existent_custom_readfunction" );
curl_setopt($oCurl, CURLOPT_INFILE,       $hReadHandle );
curl_exec($oCurl);
curl_close($oCurl);
```

This should fail.

Ideally we would have such tests for every curl option that requires a callable, and I might then try and refactor this to actually check when setting. Even though this currently means that one can define the function *after* having set the callaback...